manager/download.go

```diff
@@ -213,40 +187,120 @@ type downloader struct {
 	cfg Downloader
 
 	in *s3.GetObjectInput
-	w  io.WriterAt
+	w  io.Writer
 
 	wg sync.WaitGroup
-	m  sync.Mutex
+	m  sync.RWMutex
 
 	pos        int64
 	totalBytes int64
-	written    int64
-	err        error
+	// total bytes written to internal buffers
+	written int64
+	// flushed/written to 'w' writer
+	flushed int64
+	err     error
 
 	partBodyMaxRetries int
 }
 
+func (d *downloader) writeOutChunkIfNeeded(chunk *dlchunk) (wroteOut bool) {
+	if chunk.start < d.flushed {
+		// we received an unexpected old chunk
+		log.Printf("received unexpected old chunk starting at %d, after flushing %d", chunk.start, d.flushed)
+		return false
+	}
+
+	// if this chunk starts where we last flushed, then we are good to flush it
+	if chunk.start == d.flushed {
+		n, err := io.Copy(d.w, chunk.buf)
+		if err != nil {
+			d.setErr(err)
+		}
+		d.incrFlushed(n)
+
+		// replace/reset recycled buffer after finishing writing it
+		d.cfg.BufPool.Put(chunk.buf)
+		return true
+	}
+
+	return false
+}
+
 // download performs the implementation of the object download across ranged
 // GETs.
 func (d *downloader) download() (n int64, err error) {
 	// If range is specified fall back to single download of that range
 	// this enables the functionality of ranged gets with the downloader but
 	// at the cost of no multipart downloads.
 	if rng := aws.ToString(d.in.Range); len(rng) > 0 {
-		d.downloadRange(rng)
-		return d.written, d.err
+		return 0, errors.New("the Range option is not allowed in this downloader")
 	}
 
+	// add a write thread/queue
+	toWrite := make(chan *dlchunk, d.cfg.MaxBufferedParts)
+
+	writeWg := sync.WaitGroup{}
+	writeWg.Add(1)
+
+	go func() {
+		defer writeWg.Done()
+
+		var pendingWrites []*dlchunk
+
+		for {
+			chunk := <-toWrite
+			if chunk == nil {
+				// channel empty
+				for _, c := range pendingWrites {
+					// replace/reset recycled buffer since we are cleaning up
+					d.cfg.BufPool.Put(c.buf)
+				}
+				return
+			}
+
+			if d.getErr() != nil {
+				// clear out channel on error
+				continue
+			}
+
+			if didWrite := d.writeOutChunkIfNeeded(chunk); !didWrite {
+				// we didn't write this chunk yet, so add it to the queue of pending writes
+				pendingWrites = append(pendingWrites, chunk)
+				// TODO: would be potentially faster with a min heap but I'm lazy and there are only going to
+				// be small number of items in here (should be less than or equal to MaxBufferedParts)
+				sort.SliceStable(pendingWrites, func(i, j int) bool {
+					return pendingWrites[i].start < pendingWrites[j].start
+				})
+				continue
+			}
+
+			// we wrote out this chunk, so lets see if this unblocked any pending writes we have
+			var newPendingWrites []*dlchunk
```

Comment:
It looks to me like we're just flushing all the data from the beginning of the sorted slice that we can. Perhaps rather than iterating over all of the chunks (there won't be that many as you mentioned above), we simply flush from the beginning, retain the index of the first unwritten chunk, then re-assign the slice starting from that index?

```golang
var newSliceStart int
for i, chunk := range pendingWrites {
        if didWrite := d.writeOutChunkIfNeeded(chunk); !didWrite {
                newSliceStart = i
                break
        }
}
pendingWrites = pendingWrites[newSliceStart:]
```

Reply:
ah yeah this probably makes more sense. although trying to think over if I'd need special handling for if I wrote out all of the chunks in the `pendingWrites` I think that might end up not removing them since it'd basically do
```
pendingWrites = pendingWrites[0:]
```

I may just keep it how I have it for now since it shouldn't be too many items and I already manually tested it.